plutus-core/plutus-ir/Language/PlutusIR/Transform/Inline.hs

```diff
@@ -0,0 +1,245 @@
+code pass.
+
+Annoyingly, this requires us to redo the analysis that we do for the dead binding pass.
+TODO: merge them or figure out a way to share more work, especially since there's similar logic.
+This might mean reinventing GHC's OccAnal...
+-}
+
+processTerm
+    :: Inlining tyname name uni a term m
+    => Term tyname name uni a
+    -> m (Term tyname name uni a)
+processTerm = \case
+    v@(Var _ n) -> do
+        subst <- get
+        case lookupSubst n subst of
+            -- Not substituted for, leave it as it is
+            Nothing -> pure v
+            -- Already processed term, just put it in, don't do any further optimization here.
+            -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
+            Just (Done t) -> pure t
+    Let a NonRec bs t -> do
+        -- Process bindings, eliminating those which will be inlined unconditionally,
+        -- and accumulating the new substitutions
+        -- See Note [Removing inlined bindings]
+        bs' <- wither processSingleBinding (toList bs)
+        t' <- processTerm t
+        -- Use 'mkLet': we're using lists of bindings rather than NonEmpty since we might actually
+        -- have got rid of all of them!
+        pure $ mkLet a NonRec bs' t'
+    -- This includes recursive let terms, we don't even consider inlining them at the moment
+    t -> forMOf termSubterms t processTerm
+
+{- Note [Inlining various kinds of binding]
+We can inline term and type bindings, we can't do anything with datatype bindings.
+
+We don't actually inline type bindings at the moment, mostly because I think it
+won't get us much as they aren't created very often.
+-}
+
+processSingleBinding
+    :: Inlining tyname name uni a term m
+    => Binding tyname name uni a
+    -> m (Maybe (Binding tyname name uni a))
+processSingleBinding = \case
+    -- See Note [Inlining various kinds of binding]
+    TermBind a s v@(VarDecl _ n _) rhs -> do
+        maybeRhs' <- maybeAddSubst s n rhs
+        case maybeRhs' of
+            Just rhs' -> pure $ Just $ TermBind a s v rhs'
+            Nothing -> pure Nothing
+    -- Not a strict binding, just process all the subterms
+    b -> Just <$> forMOf bindingSubterms b processTerm
+
+maybeAddSubst
+    :: Inlining tyname name uni a term m
+    => Strictness
+    -> name
+    -> Term tyname name uni a
+    -> m (Maybe (Term tyname name uni a))
+maybeAddSubst s n rhs = do
+    -- Only do PostInlineUnconditional
+    -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
+    rhs' <- processTerm rhs
+    doInline <- postInlineUnconditional s rhs'
+    if doInline then do
+        modify (\subst -> extendSubst n (Done rhs') subst)
+        pure Nothing
+    else pure $ Just rhs'
+
+{- Note [Inlining criteria]
+What gets inlined? We don't really care about performance here, so we're really just
+angling to simplify the code without making things worse.
+
```

Comment:
"We don't really care about performance here".  Is that true?  One of the things this will be doing for us is removing the indirection in calls to builtins, and I think (or at least hope) that that will make execution noticeably more efficient.